src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SslStreamSecurityUpgradeProvider.cs

```diff
@@ -566,16 +414,210 @@ internal override void Close(TimeSpan timeout)
         protected override Stream OnInitiateUpgrade(Stream stream, out SecurityMessageProperty remoteSecurity)
         {
             OutWrapper<SecurityMessageProperty> remoteSecurityWrapper = new OutWrapper<SecurityMessageProperty>(); 
-            Stream retVal = this.OnInitiateUpgradeAsync(stream, remoteSecurityWrapper).GetAwaiter().GetResult();
+            Stream retVal = OnInitiateUpgradeAsync(stream, remoteSecurityWrapper).GetAwaiter().GetResult();
             remoteSecurity = remoteSecurityWrapper.Value; 
 
             return retVal; 
         }
 
 #if FEATURE_NETNATIVE
-        protected override Task<Stream> OnInitiateUpgradeAsync(Stream stream, OutWrapper<SecurityMessageProperty> remoteSecurityWrapper)
+        protected override async Task<Stream> OnInitiateUpgradeAsync(Stream stream, OutWrapper<SecurityMessageProperty> remoteSecurityWrapper)
         {
-            throw ExceptionHelper.PlatformNotSupported("SslStreamSecurityUpgradeInitiator.InInitiateUpgradeAsync");
+            if (WcfEventSource.Instance.SslOnInitiateUpgradeIsEnabled())
+            {
+                WcfEventSource.Instance.SslOnInitiateUpgrade();
+            }
+
+            // There is currently no way to convert a .Net X509Certificate2 to a UWP Certificate. The client certificate
+            // needs to be provided by looking it up in the certificate store. E.g.
+            //
+            //     factory.Credentials.ClientCertificate.SetCertificate(
+            //         StoreLocation.CurrentUser,
+            //         StoreName.My,
+            //         X509FindType.FindByThumbprint,
+            //         clientCertThumb);
+            //
+            // The certificate is retrieved using .Net api's and UWP api's. An artifical X509Extension is used to attach the
+            // UWP certificate to the .Net X509Certificate2. This is then retrieved at the point of usage to use with UWP
+            // networking api's.
+
+            Certificate clientCertificate = null;
+
+            if (_clientToken != null)
+            {
+                foreach (var extension in _clientToken.Certificate.Extensions)
+                {
+                    var attachmentExtension =
+                        extension as X509CertificateInitiatorClientCredential.X509UwpCertificateAttachmentExtension;
+                    if (attachmentExtension != null && attachmentExtension.AttachedCertificate != null)
+                    {
+                        clientCertificate = attachmentExtension.AttachedCertificate;
+                        break;
```

Comment:
In the ClientCertificate.SetCertificate method, we throw if we don't find exactly 1 certificate.